POEM_056.md

```diff
@@ -0,0 +1,349 @@
+we also need to associate output names with function return values. Those return values, if they are 
+simple variables, for example, `return x, y`, will give us the output variable names we need.  
+But in those cases where the function returns expressions rather than simple variables, we need 
+another way to specify what the names of those output variables should be.  The `output_names` 
+method provides a concise way to do this, for example:
+
+
+```python
+def func(a, b, c):
+    return a * b * c, a * b -c
+
+f = omf.wrap(func).output_names('d', 'e')  # name of return values are 'd' and 'e'
+```
+
+If we don't want to bother with a separate decorator for output names, we could instead use the
+`add_outputs` decorator mentioned earlier, for example:
+
+```python
+def func(a, b, c):
+    return a * b * c, a * b -c
+
+# names of return values are 'd' and 'e' and they have no other metadata
+f = omf.wrap(func).add_outputs(d={}, e={})  
+```
+
+As mentioned above, if the function's return values are simple variable names, we don't need to
+call `output_names` because we can determine the names from inspecting the function, e.g., 
+
+
+```python
+def func(a, b, c):
+    d = a * b * c
+    e = a * b -c
+    return d, e  # output names are 'd' and 'e'. no call to output_names needed
+```
+
+Note that if neither `output_names` nor `add_outputs` is specified and the output names cannot be 
+determined by inspection of the return values, then they must be specified using `add_output` calls, 
+and the order of those `add_output` calls determines how those names map to the return value positions.  For example:
+
+```python
+def func(x):
+    return x.dot(np.random.random(2))., x*1.5
+
+f = (omf.wrap(func)
+        .add_input('x', shape=(2,2))
+        .add_output('y', shape=2)
+        .add_output('z', shape=(2,2))) 
+```
+
+In the example above, the output names would be assumed to be `['y', 'z']`.
+
+
+### Getting variable names
+
+Lists of input names and output names can be retrieved by calling `get_input_names` and 
+`get_output_names` respectively, e.g., 
+
+```python
+
+invar_names = f.get_input_names()
+outvar_names = f.get_output_names()
+
+```
+
+## Partial derivatives
+
+### Setting partial derivative information
+
+Metadata that will help OpenMDAO or potentially other libraries to compute partial derivatives
+for the function can be defined using the `declare_partials` and `declare_coloring` methods.
+For example:
+
+```python
+
+def func(x, y, z=3): 
+    foo = np.log(z)/(3*x+2*y)
+    bar = 2*x+y
+    return foo, bar
+
+f = (omf.wrap(func)
+        .declare_partials(of='*', wrt='*', method='cs')
+        .declare_coloring(wrt='*', method='cs')
+        .defaults(shape=4))
+```
+
+The args for the `declare_partials` and `declare_coloring` methods match those
+of the  `declare_partials` and `declare_coloring` methods of an OpenMDAO component.  Multiple calls
+can be made to either decorator to set up different partials/colorings.
```

Comment:
is that true in regular om with respect to coloring? Can you color some stuff with CS and others with FD? I thought we made it so you had to pick one or the other. 

I know its true for partials. Just not sure about coloring.

Reply:
You're right.  Looks like we don't support multiple calls to declare_coloring for a given component.

Reply:
I cleared up the explanation of this.